torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -373,92 +364,339 @@ class _ExecOrderWarnStatus(Enum):
+            return None
+        target_index = current_index + 1
+        if target_index >= len(self.handles_pre_forward_order):
+            return None
+        target_handles_key = self.handles_pre_forward_order[target_index]
+        return target_handles_key
 
-    def get_unflat_param_names(self, param_index: int) -> List[str]:
-        """Returns a :class:`list` of unflattened parameter names comprising
-        the flattened parameter with index ``param_index`` or an empty
-        :class:`list` if ``param_index`` is invalid."""
-        param = self.get_param(param_index)
-        if param is None:
-            return []
-        assert param in self._param_to_unflat_param_names, \
-            "Internal data structures out of sync; check `init()`"
-        return self._param_to_unflat_param_names[param]
+    def get_handles_to_backward_prefetch(
+        self,
+        current_handles_key: Tuple[FlatParamHandle, ...],
+    ) -> Optional[Tuple[FlatParamHandle, ...]]:
+        """
+        Returns the handles key of the handles to backward prefetch given the
+        current handles key or ``None`` if there is no valid handles key to
+        prefetch.
+        """
+        current_index = self.handles_to_post_forward_order_index.get(current_handles_key, None)
+        if current_index is None:
+            return None
+        target_index = current_index - 1
+        if target_index < 0:
+            return None
+        target_handles_key = self.handles_post_forward_order[target_index]
+        return target_handles_key
+
+    def record_post_forward(self, handles: List[FlatParamHandle]) -> None:
+        """
+        Records ``handles`` in the post-forward order, where ``handles`` should
+        be a group of handles used in the same module's forward. If ``handles``
+        is empty, then it is omitted.
+
+        Unlike :meth:`record_pre_forward`, this records the order *every*
+        iteration with the expectation that the recorded order is reset in
+        :meth:`next_iter`.
+        """
+        if not handles:
+            return
+        handles_key = tuple(handles)
+        if handles_key in self.handles_to_post_forward_order_index:
+            return
+        index = len(self.handles_post_forward_order)
+        self.handles_to_post_forward_order_index[handles_key] = index
+        self.handles_post_forward_order.append(handles_key)
+
+    def record_pre_forward(self, handles: List[FlatParamHandle]) -> None:
+        """
+        Records ``handles`` in the pre-forward order on the first iteration,
+        where ``handles`` should be a group of handles used in the same
+        module's forward. If ``handles`` is empty, then it is omitted.
+
+        If the distributed debug level is at least INFO, then this additionally
+        checks the execution order across ranks. See :meth:`_check_order` for
+        details.
+        """
+        # TODO (awgu): For now, we exclude modules with no parameters from the
```

Comment:
We may actually want to include the empty handles keys. I am looking into it right now.

Reply:
I re-included the "bubbles", but the prefetching implementation still has some slight semantic difference compared to the existing one. This PR's implementation will not have a module with no parameters prefetch for the next layer. In order to fix this/preserve the existing semantics, I need to change the key to be maybe `Tuple[nn.Module, Tuple[FlatParamHandle, ...]]` (i.e. include the module in the key) or else we cannot differentiate empty handles keys (`(,)`).